File: cpp/src/Test7.cpp
#include "TestHelpers.h"
#include "Problem7.h"
#include <cstdio>
#include <iostream>
#include <fstream>
#include <vector>
#include <string>

class Test7 {
private:
static std::vector<int> getData(std::string& f) {
    std::string dataFileName = TestHelpers::getDataFile(f);
    std::vector<int> res;
    int num_values, i, x;

    std::ifstream ifs(dataFileName);

    if (ifs) {
        ifs >> num_values >> i >> x;
        res.push_back(i);
        res.push_back(x);
    } else {
        std::cout << dataFileName << " failed to be read" << std::endl;;
        return std::vector<int>();
    }

    for (int i = 0; i < num_values; ++i) {
        int temp;
        ifs >> temp;
        res.push_back(temp);
    }
    
    if (res.size() != num_values + 2) {
        std::cout << "Format error: Actual and expected number of values do not match: " << dataFileName << std::endl;;
        exit(1);
    }

    return res;
}

public:
static bool runTest(std::string f) {
    std::vector<int> data = getData(f);
    int i = data[0], x = data[1];
    std::vector<int> values = {};
    for (size_t j = 2; j < data.size(); ++j) {
        values.push_back(data[j]);
    }
    Problem7::fill(values, i, x);
    std::vector<int> expected = TestHelpers::getResult(f);


    if (expected.size() == 0) {
        return false;
    }
    
    if (values.size() == 0) {
        std::cout << "Failed: actual result is null\n";
	return false;
    } else if (values == expected) {
        return true;
    } else {
        std::cout << "Failed actual result and expected result do not match\n";
        printDiff(values, expected);
	return false;
    }
}

static void printDiff(const std::vector<int>& actual, const std::vector<int>& expected) {
    if (actual.size() != expected.size()) {
        std::cout << "Actual length (" << actual.size() << ") and expected length (";
        std::cout << expected.size() << ") do not match\n";
        return;
    }

    for (int i = 0; i < actual.size(); ++i) {
        if (actual[i] != expected[i]) {
            std::cout << "  Actual and expected results do not match at index: " << i << std::endl;
            std::cout << "    actual @ index " << i << " is " << actual[i] << std::endl;
            std::cout << "    expected @ index " << i << " is " << expected[i] << std::endl;
            return;
        }
    }
    return;
}
};

int main (int argc, char* argv[]) {
    std::vector<std::string> files = TestHelpers::getFiles(argc, argv);
    if (files.size() == 0) {
        std::cout << "no test file found\n"; 
        exit(1);
    } 

    int count = 0;
    for (size_t i = 0; i < files.size(); ++i) {
        std::string f = files[i];
	std::cout << "Test " << f << ": ";	 	
        if (Test7::runTest(f)) {
	  std::cout << "PASSED\n";	  
	  count++;
	} else {
	  std::cout << "FAILED\n";	  
	}
    }

    if (count == files.size()) {
      return 42;
    } else {
      return 0;
    }
}

File: java/src/Test7.java
import java.util.Scanner;
import java.util.Arrays;
import java.util.List;
import java.util.ArrayList;
import java.io.*;
import java.util.NoSuchElementException;

public class Test7 extends Test {
    public boolean runTest(File f, boolean verbose) {
	String inputFilename = f.getPath();
	Scanner scanner = null;
	try {
	    scanner = new Scanner(f);
	} catch (FileNotFoundException e) {
	    if (verbose) {
		System.out.println("Results file not found:" + inputFilename);
	    }
	    return false;
	}

	// Format: number of values, index, replacement value
	List<Integer> params = TestHelpers.getIntegerListFromFile(scanner, 3, inputFilename);
	int[] values = TestHelpers.convertIntegerListToArray(TestHelpers.getIntegerListFromFile(scanner, params.get(0), inputFilename));
	Problem7.fill(values, params.get(1), params.get(2));
	File resultsFile = TestHelpers.getResultsFile(f);
	int[] expected = TestHelpers.convertIntegerListToArray(TestHelpers.getIntegerListFromFile(resultsFile, values.length));

	if (expected == null) {
	    if (verbose) {
		System.out.println("Bad results file:" + resultsFile.getPath());
	    }
	    // bad results file.
	    return false;
	}

	if (values == null) {
	    if (verbose) {
		System.out.println("Failed: actual result is null");
	    }
	    return false;
	} else {
	    for (int i = 0; i < values.length; i++) {
		if (values[i] != expected[i]) {
		    if (verbose) {
			System.out.println(String.format("  Actual and expected results do not match at index: %d", i));
			System.out.println(String.format("    actual @ index %d is %d", i, values[i]));
			System.out.println(String.format("    expected @ index %d is %d", i, expected[i]));
		    }
		    return false;
		}
	    }
	    return true;
	}
    }


    public static void main(String[] args) {
	System.exit(TestHelpers.doTests(args, new Test7()));	
    }
}
